### ISSUE TITLE
Consider making AsyncResult its own variant type rather than piggybacking on AsyncData

### ISSUE THREAD

**Issue by andywhite37** [reazen/relude#77]
`AsyncData` represents data that can be in 4 loading states: `Init`, `Loading`, `Reloading(a)`, and `Complete(a)`.  There is no way to represent failed loading with this type.

`AsyncResult` was added as a wrapper for `AsyncData.t(Result.t('a, 'e))` to allow loading errors, like `Complete(Error("Failed"))`.

There is currently an awkward case where if `Reloading` fails, you no longer have access to the value that you had before, which was kind of the point of adding the `Reloading` case.  One example scenario where this is awkward is if you are doing a form PUT, and using the `Reloading` state, and it fails, you lose the entire context for rendering the (previous values of the) form.

Maybe we should not make AsyncResult an alias of `AsyncData` with an inner  `Result`, and just make it it's own variant like:

```re
type t('a, 'e) =
  | Init
  | Loading
  | Reloading('a)
  | ReloadingError('a, 'e) // This is the special case with (previous) value and an error
  | Complete('a)
  | Error('e);
```

We'd probably want to use slightly different constructor names than `AsyncData` to avoid clashes.  If people are using the functions for dealing with `AsyncResult`, this shouldn't cause too many problems, but it would cause breaking changes for people matching on the constructors.

@mlms13 - thoughts?

**Comment by andywhite37** [comment]
My current workaround is to make the error type a `(option('a), 'e)`, like `Relude.AsyncResult.t('a, (option('a), 'e)`.  This way - I can optionally have a value along with the error.

**Comment by andywhite37** [comment]
I was thinking that this is kind of like an `AsyncIor`, with the case that has both a success value and an error value.  I might play around with that as a concept, so as not to break `AsyncResult`, which is used all over the place.

`Ior` (inclusive or) looks like this (we have it in Relude as `ResultIor`):

```
type t('a, 'e) = | IOk('a) | IError('e) | IBoth('a, 'e)
```

The purpose of this is to allow for doing applicative validation, but with the ability to pickup both errors and successes at the same time, and combine all the errors at the end.

**Comment by erlandsona** [comment]
What we need for situations like this is probably just to shamelessly steal from [Topo's Smash lib](https://hackage.haskell.org/package/smash-0.1.1.0/docs/src/Data.Can.html#line-538)

**Comment by jihchi** [comment]
`Ior` (we have it in Relude as `ResultIor`):

```re
type t('a, 'e) =
  | IOk('a)
  | IError('e)
  | IBoth('a, 'e);
```

`Data.Can` from [Topo's Smash lib](https://hackage.haskell.org/package/smash-0.1.1.0/docs/Data-Can.html):

```re
type t('a, 'b) =
  | Non
  | One('a)
  | Eno('b)
  | Two('a, 'b);
```

Interesting

**Comment by andywhite37** [comment]
Yeah, I was going to say that we have the `Ior` type (which used to be `ResultIor`, but is now just `Ior`, which I recently changed to lose the `Ok`/`Error` semantics.

It looks like this now:

```
type t('a, 'b) =
  | This('a)
  | That('b)
  | Both('a, 'b);
```

This was based on the [scala cats `Ior`](https://typelevel.org/cats/datatypes/ior.html), and the [haskell/purescript `These`](https://hackage.haskell.org/package/these-1.0.1/docs/Data-These.html) types.

**Comment by erlandsona** [comment]
I'm workin' through a transcription of it from Haskell -> Reason now. I'll PR back here when I'm done. But yeh I suppose I could just use `option(Ior.t('a, 'e))` and mix & match the functions from each module to get at what I need. 
I think having it wrapped up in a single type makes workin' with it more intuitive maybe?

**Comment by andywhite37** [comment]
Yeah, I think a separate type makes sense, like the `Data.Can` type. The nested types always get clunky to work with, and this is general enough to warrant it's own implementation.
